EKF/control.cpp

```diff
@@ -158,10 +158,85 @@ void Ekf::controlFusionModes()
 		}
 	}
 
-	// Handle the case where we have rejected height measurements for an extended period
-	// This excessive vibration levels can cause this so a reset gives the filter a chance to recover
-	// After 10 seconds without aiding we reset to the height measurement
-	if (_time_last_imu - _time_last_hgt_fuse > 10e6) {
+	/*
+	 * Handle the case where we have not fused height measurements recently and
+	 * uncertainty exceeds the max allowable. Reset using the best available height
+	 * measurement source, continue using it after the reset and declare the current
+	 * source failed if we have switched.
+	*/
+	if ((P[8][8] > sq(_params.hgt_reset_lim)) && ((_time_last_imu - _time_last_hgt_fuse) > 5e6)) {
+		// handle the case where we are using baro for height
+		if (_control_status.flags.baro_hgt) {
+			// check if GPS height is available
+			gpsSample gps_init = _gps_buffer.get_newest();
+			bool gps_hgt_available = ((_time_last_imu - gps_init.time_us) < 2 * GPS_MAX_INTERVAL);
```

Comment:
Why do you add a factor of 2 before the MAX_INTERVAL. Isn't the max already saying that it is a threshold?